modules/db/src/main/scala/gem/dao/UserDao.scala

```diff
@@ -11,55 +11,74 @@ import doobie.imports._
-    Statements.selectWithRoles(id, pass)
-      .list
-      .map { rows =>
-        rows.headOption.map {
-          case (i, f, l, e, s, _, _) =>
-            val map = rows.collect {
-              case (_, _, _, _, _, Some(pid), Some(role)) => Map(pid -> Set(role))
-            }.suml
-            User(i, f, l, e, s, map)
-        }
-      }
+  /**
+   * Select the root user, for system processes. Always succceds if the system is consistent;
+   * raises an exception if the root user is missing.
+   * @group Queries
+   */
+  val selectRootUser: ConnectionIO[User[ProgramRole]] =
+    OptionT(selectUser(User.Id.Root)).getOrElse(sys.error("No root user. Cannot continue."))
+
+  // Helper method to add roles to a selected user, if any.
+  private def selectUserImpl(q: Query0[User[Nothing]]): ConnectionIO[Option[User[ProgramRole]]] = {
+    for {
+      u <- OptionT(q.option)
+      r <- selectRoles(u.id).liftM[OptionT]
+    } yield u.copy(roles = r)
+  } .run
+
+  /**
+   * Select the given user, if any, with roles.
+   * @group Queries
+   */
+  def selectUser(uid: User.Id): ConnectionIO[Option[User[ProgramRole]]] =
+    selectUserImpl(Statements.selectUser(uid))
+
+  /**
+   * Select the given user, if any, with roles, if the id and password match a known user.
+   * @group Queries
+   */
+  def selectUserʹ(uid: User.Id, password: String): ConnectionIO[Option[User[ProgramRole]]] =
+    selectUserImpl(Statements.selectUserʹ(uid, password))
+
+  /**
+   * Select the map of roles associated with the given user.
+   * @group Queries
+   */
+  def selectRoles(id: User.Id): ConnectionIO[Map[Program.Id, Set[ProgramRole]]] =
+    Statements.selectRoles(id).list.map(_.foldMap { case (k, v) => Map((k -> Set(v))) })
 
+  /**
+   * Attempts to change the specified user's password, yielding success. A cause of failure (user
+   * not found, old password incorrect) is not given.
+   * @group Updates
+   */
   def changePassword(uid: User.Id, oldPassword: String, newPassword: String): ConnectionIO[Boolean] =
     Statements.changePassword(uid, oldPassword, newPassword).run.map(_ === 1)
 
   object Statements {
 
-    def select(id: String): Query0[User[Nothing]] =
-      sql"""
+    private def selectUserImpl(id: String, and: Fragment): Query0[User[Nothing]] =
+      (fr"""
         SELECT id, first, last, email, staff
         FROM gem_user
         WHERE id = $id
-      """.query[(String, String, String, String, Boolean)]
-         .map { case (i, f, l, e, s) =>
-           User(i, f, l, e, s, Map.empty)
-         }
+       """ ++ and).query[(String, String, String, String, Boolean)].map {
+        case (i, f, l, e, s) => User(i, f, l, e, s, Map.empty)
+      }
+
+    def selectUser(id: String): Query0[User[Nothing]] =
+      selectUserImpl(id, Fragment.empty)
+
+    def selectUserʹ(id: String, password: String): Query0[User[Nothing]] =
+      selectUserImpl(id, fr"AND md5 = md5($password)")
```

Comment:
I think in general any standard hashing algorithm is not recommended for storing passwords without several safety measures like salting, authenticating, etc.

I'd use one of the specialized libraries to store passwords `bcrypt` `scrypt` or `PKDF2` and forget about the details. I'm sure there'll be plenty of implementations in Java/Scala.

If we don't want to bother evaluating crypto algorithms every time we need to pick one, I'd use all the defaults in the [libsodium](https://download.libsodium.org/doc/) and forget about it. I never had the chance to use it because of legacy constraints but here I don't we have that problem.